cc/main.cc

```diff
@@ -757,40 +683,63 @@ void Puzzle() {
   ParseMctsPlayerOptionsFromFlags(&options);
   options.verbose = false;
 
-  using Pair = std::pair<std::unique_ptr<MctsPlayer>, Move>;
-  std::vector<Pair> puzzles;
-  for (const auto& moves : games) {
-    std::vector<std::unique_ptr<MctsPlayer>> players(moves.size());
-    for (auto& player : players) {
-      player.reset(new MctsPlayer(factory->NewDualNet(FLAGS_model), options));
+  std::atomic<size_t> total_moves(0);
+  std::atomic<size_t> correct_moves(0);
+
+  std::vector<std::thread> threads;
+  std::vector<std::string> basenames;
+  MG_CHECK(file::ListDir(FLAGS_sgf_dir, &basenames));
+  for (const auto& basename : basenames) {
+    if (!absl::EndsWith(basename, ".sgf")) {
+      continue;
     }
-    for (const auto& move : moves) {
-      puzzles.emplace_back(std::move(players.back()), move);
-      players.pop_back();
-      for (auto& player : players) {
-        MG_CHECK(player->PlayMove(move.c));
+    threads.emplace_back([&]() {
+      // Read the main line from the SGF.
+      auto path = file::JoinPath(FLAGS_sgf_dir, basename);
+      std::string contents;
+      MG_CHECK(file::ReadFile(path, &contents));
+      sgf::Ast ast;
+      MG_CHECK(ast.Parse(contents));
+      auto trees = GetTrees(ast);
+      MG_CHECK(!trees.empty());
+      auto moves = trees[0]->ExtractMainLine();
+
+      total_moves += moves.size();
+
+      // Create player.
+      auto player = absl::make_unique<MctsPlayer>(
+          model_factory->NewDualNet(FLAGS_model), options);
+      model_factory->StartGame(player->network(), player->network());
```

Comment:
maybe puzzlers can have a single-network at some point in the future?  I dunno.  Most of the ad-hoc stuff with nets should be done in python anyway.